### DISCUSSION TITLE
[Extensibility + Web Parts] Building an extensibility library on SPFx 1.19.0 or 1.20.0 causes run-time errors in the Search Web Part's LayoutHelper function "getUsedSlotNames"

### DISCUSSION THREAD

**Discussion by tandrasi** [microsoft-search/pnp-modern-search#4047] [Q&A]
**Version used**
PnP Modern Search - Search Web Parts - v4.13.1.0
@pnp/modern-search-extensibility - v1.19.0 or v1.18.2 (read more below)
SPFx - v1.19.0 or v1.20.0

**Describe the bug**
We have a solution dedicated to various SPFx libraries, such as the Search Extensibility library. The entire solution is built on SPFx 1.18.2. We are doing a sweep to upgrade existing projects to 1.19.0/1.20.0 to future proof them. Existing internal frameworks and projects which rely on them have been upgraded successfully to ensure they work with Webpack 5 under SPFx 1.20.0.

However, attempting to upgrade our SPFx libraries solution to 1.19.0, without touching any other code, has led to a runtime (NOT build) error with the Search Web Parts. It fails at the **getUsedSlotNames** function in LayoutHelper.ts, specifically at line 276 as "templateContent" is being passed a "Module" object rather than a string, causing the "trim()" function to fail. This causes the custom templates to not load and the entire web part to break.

**To Reproduce**
Detailed steps to reproduce the behavior:
1. Set Search Results web part to use a custom search extensibility template built on SPFx 1.19.0 or 1.20.0 (read more in **Additional Context);**
2. Failure :(

**Expected behavior**
I expect this run-time issue not to occur, just like in 1.18.2.

**Screenshots**
Our "libraries" solution has 3 layouts which we have been maintaining and using without a hitch, all with very complex rendering baked in. All this is working in SPFx 1.18.2:
![image](https://github.com/user-attachments/assets/15d8c033-4c80-4f87-a25f-e0a03b2abec0)

Upgrading to 1.19.0 leaves default templates working as getUsedSlotNames's "templateContent" parameter correctly gets passed as a string:
![image](https://github.com/user-attachments/assets/7c9449e8-0340-45a6-9622-3a28284a4d9a)

However, upgrading to 1.19.0 causes custom templates to fail as getUsedSlotNames's "templateContent" parameter incorrectly gets passed as an object (a "Module", to be exact) and therefore the code can't call "trim":
![image](https://github.com/user-attachments/assets/d4a554be-f436-4cb8-a454-e43c3ffe5f88)

![image](https://github.com/user-attachments/assets/d5a579a1-5353-4c7a-8f40-a2c7ee3d0e68)

![image](https://github.com/user-attachments/assets/c0c5f966-c62e-454a-b4b3-d520d49ee077)

**Desktop (please complete the following information):**
 - Browser: Chromium (Chrome, Edge - latest)

**Additional context**

Things I have tried:
1. **fails** Using SPFx 1.19.0 with @pnp/modern-search-extensibility@1.19.0.
2. **fails** Using SPFx 1.20.0 with @pnp/modern-search-extensibility@1.19.0 (with "any" casting to circumvent build errors).
3. **fails** Using SPFx 1.19.0 with @pnp/modern-search-extensibility@1.18.2  (with "any" casting to circumvent build errors).
4. **works** Using SPFx 1.18.2 with @pnp/modern-search-extensibility@1.18.2

Theories:
I think this is due to SPFx 1.19.0+'s Webpack 5 builds. It has drastically changed how we need **import** packages and include pollyfills. 

I created this original issue with SPFx 1.19.0 as our builds were failing but we didn't know why with [Object][object] errors: 
https://github.com/SharePoint/sp-dev-docs/issues/9675. This was closed in favour of a more descriptive issue: https://github.com/SharePoint/sp-dev-docs/issues/9834. 

The SPFx 1.20.0 release address these issues, with it being referenced directly in the changelog:

![image](https://github.com/user-attachments/assets/e2f4b790-c9c4-4c02-8052-17c772f1df2a)

**Comment by wobba** [comment]
`v4.13.1.0`of the web parts should use use `v1.18.2` of extensibility https://www.npmjs.com/package/@pnp/modern-search-extensibility/v/1.18.2. I had issues updating to SPFX v1.19/20 so will stay at 1.18.2 for the time being. Unfortunately loading of libs like this require all parts to be on the same SPFx version.

If someone takes the time to resolve v1.19 issues then we can move ahead.

**Reply by tandrasi** [reply]
Thanks @wobba!

**Reply by tandrasi** [reply]
@wobba, hi again! Listen, so I've just deployed version 4.16.0 of PnP Modern Search to our tenancy. I've also upgraded our internal "library" solution to SPFx 1.20.0 and we're using v1.20.1 of the search extensibility framework. All SPFx versions are now matching at "1.20.0".

However, I still get the error as mentioned above:

![image](https://github.com/user-attachments/assets/24f07864-d306-4792-9e40-46d3d19a715f)

I have even tried changing the names of our templates to remove any special characters, but nothing. I've also tried redeploying our "library" solution with the latest versions, but nothing.

Any recommendations or suggestions?

**Reply by tandrasi** [reply]
@wobba I have re-created my previous bug here: https://github.com/microsoft-search/pnp-modern-search/issues/4237.

**Reply by jblueml** [reply]
Just posted a possible solution on this in #4237
